ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -1674,172 +1672,187 @@ extension SignalProducerProtocol {
-				// have terminated.
-				lifetime = nil
-			}
-
-			if shouldStartUnderlyingProducer {
-				self.take(during: lifetime!)
-					.start(initializedObserver)
-			}
-		}
-	}
-}
-
-extension SignalProducer {
-	private static func bufferingProducer(upTo capacity: Int) -> (SignalProducer, Signal<Value, Error>.Observer) {
-		precondition(capacity >= 0, "Invalid capacity: \(capacity)")
-
-		// Used as an atomic variable so we can remove observers without needing
-		// to run on a serial queue.
-		let state: Atomic<BufferState<Value, Error>> = Atomic(BufferState())
-
-		let producer = self.init { observer, disposable in
-			// Assigned to when replay() is invoked synchronously below.
-			var token: RemovalToken?
-
-			let replayBuffer = ReplayBuffer<Value>()
-			var replayValues: [Value] = []
-			var replayToken: RemovalToken?
-			var next = state.modify { state in
-				replayValues = state.values
-				if replayValues.isEmpty {
-					token = state.observers?.insert(observer)
-				} else {
-					replayToken = state.replayBuffers.insert(replayBuffer)
+		let bootstrap = ActionDisposable {
+			// Start the underlying producer.
+			self.take(during: lifetime)
+				.start { event in
+					let originalState = state.modify { state in
+						state.enqueue(event)
+					}
+					originalState.broadcast(event)
 				}
-			}
+		}
 
-			while !replayValues.isEmpty {
-				replayValues.forEach(observer.sendNext)
+		return SignalProducer { observer, disposable in
+			// Don't dispose of the original producer until all observers
+			// have terminated.
+			disposable += { _ = lifetime }
+
+			repeat {
+				do {
+					var token: RemovalToken?
+					try state.modify { state in
+						token = try state.observe(observer)
+					}
 
-				next = state.modify { state in
-					replayValues = replayBuffer.values
-					replayBuffer.values = []
-					if replayValues.isEmpty {
-						if let replayToken = replayToken {
-							state.replayBuffers.remove(using: replayToken)
+					if let token = token {
+						disposable += {
+							state.modify { state in
+								state.removeObserver(using: token)
+							}
 						}
-						token = state.observers?.insert(observer)
 					}
-				}
-			}
-
-			if let terminationEvent = next.terminationEvent {
-				observer.action(terminationEvent)
-			}
 
-			if let token = token {
-				disposable += {
-					state.modify { state in
-						state.observers?.remove(using: token)
-					}
-				}
-			}
-		}
+					// Start the underlying producer if it has never been started.
+					bootstrap.dispose()
 
-		let bufferingObserver: Signal<Value, Error>.Observer = Observer { event in
-			let originalState = state.modify { state in
-				if let value = event.value {
-					state.add(value, upTo: capacity)
-				} else {
-					// Disconnect all observers and prevent future
-					// attachments.
-					state.terminationEvent = event
-					state.observers = nil
+					// Terminate the replay loop.
+					return
+				} catch ReplayError<Value>.pending(let values) {
```

Comment:
The compiler somehow does not permit `catch let ReplayError<Value>.pending(values)`.